client/macro_processor.py

```diff
@@ -0,0 +1,141 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import fnmatch
+import os
+import re
+import shutil
+import yaml
+
+
+from argparse import Namespace
+from yaml.loader import SafeLoader
+from os.path import dirname, isfile, join
+from typing import Dict
+
+
+class MacrosProcessor:
+    """A processor to handle macros in the query files during the pre-processing and post-processing stages of a Batch
+    Sql Translation job.
+
+    """
+    def __init__(self, context: Namespace):
+        self.context = context
+        if context.macro_map:
+            self.enable_macro_substitution = True
+            self.macros_subst = MapBasedExpander(context.macro_map)
+        else:
+            self.enable_macro_substitution = False
+
+    def pre_process(self, input_dir: str, output_dir: str):
+        assert self.enable_macro_substitution, "Macros processing not enabled."
+        self.process_inputs(input_dir, output_dir)
+
+    def post_process(self, input_dir: str, output_dir: str):
+        assert self.enable_macro_substitution, "Macros processing not enabled."
+        self.process_inputs(input_dir, output_dir, use_reversed_map=True)
+
+    def process_inputs(self, input_dir: str, output_dir: str, use_reversed_map=False):
+        """Replaces macros with values for every file in the input folder and save outputs in a new folder.  Macro
+        replacement doesn't apply for files with extension of .zip, .json, and .csv.
+
+        Args:
+            input_dir: path to the input directory.
+            output_dir: path to the output directory.
+            use_reversed_map: whether to use the reversed macro_replacement_maps.
+        """
+        for file in os.listdir(input_dir):
+            file_path = join(input_dir, file)
+            if isfile(file_path):
+                if not file.lower().endswith(('.zip', '.json', '.csv')) and not file.startswith("."):
+                    print("Applying token maps to file %s" % file_path)
+                    self.process_file(file_path, input_dir, output_dir, use_reversed_map)
+                elif not file.startswith("."):
+                    shutil.copy(file_path, file_path.replace(input_dir, output_dir))
+
+    def process_file(self, file_path: str, input_dir: str, output_dir: str, use_reversed_map):
+        """Applies the macro maps to a specific file and saves the output file in the output directory with the same
+        subdirectory and name as the input file.
+        """
+        assert isfile(file_path), "Can't find a file at \"%s\"." % file_path
+
+        # Prepare the output file path.
+        output_file_path = file_path.replace(input_dir, output_dir)
```

Comment:
This is dangerous. If we know the input_dir is a prefix, we should remove the prefix. If it's relative, we might end up replacing multiple occurrences when we only mean to replace the one at index 0

Reply:
Done.  I also tested this at different config cases, e.g., relative vs. absolute path, sub-dir in the input folder.